pkg/inframetadata/internal/hostmap/hostmap.go

```diff
@@ -0,0 +1,143 @@
+
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata/gohai"
+	"github.com/DataDog/opentelemetry-mapping-go/pkg/inframetadata/payload"
+)
+
+// HostMap maps from hostnames to host metadata payloads.
+type HostMap struct {
+	// mu is the mutex for the host map and updater.
+	mu sync.Mutex
+	// hosts map
+	hosts map[string]payload.HostMetadata
+}
+
+// New creates a new HostMap.
+func New() (*HostMap, error) {
+	return &HostMap{
+		mu:    sync.Mutex{},
+		hosts: make(map[string]payload.HostMetadata),
+	}, nil
+}
+
+func getStrField(m pcommon.Map, key string) (string, bool, error) {
+	val, ok := m.Get(key)
+	if !ok {
+		// Field not available, don't update but don't fail either
+		return "", false, nil
+	}
+
+	if val.Type() != pcommon.ValueTypeStr {
+		return "", false, fmt.Errorf("%q has type %q, expected type \"Str\" instead", key, val.Type())
+	}
+
+	return val.Str(), true, nil
+}
+
+func isAWS(m pcommon.Map) (bool, error) {
+	cloudProvider, ok, err := getStrField(m, conventions.AttributeCloudProvider)
+	if err != nil {
+		return false, err
+	} else if !ok {
+		// no cloud provider field
+		return false, nil
+	}
+	return cloudProvider == conventions.AttributeCloudProviderAWS, nil
+}
+
+func getInstanceID(m pcommon.Map) (string, bool, error) {
+	if onAWS, err := isAWS(m); err != nil || !onAWS {
+		return "", onAWS, err
+	}
+	return getStrField(m, conventions.AttributeHostID)
+}
+
+func getEC2Hostname(m pcommon.Map) (string, bool, error) {
+	if onAWS, err := isAWS(m); err != nil || !onAWS {
+		return "", onAWS, err
+	}
+	return getStrField(m, conventions.AttributeHostName)
+}
+
+// Update the information about a given host by providing a resource.
+// The function reports:
+//   - Whether the information about the `host` has changed
+//   - Any non-fatal errors that may have occurred during the update
+//
+// Partial modifications will still be applied even with non-fatal errors.
+func (m *HostMap) Update(host string, res pcommon.Resource) (changed bool, err error) {
+	md := payload.HostMetadata{
+		Flavor:  "otelcol-contrib",
+		Meta:    &payload.Meta{},
+		Tags:    &payload.HostTags{},
+		Payload: gohai.NewEmpty(),
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	var found bool
+	if old, ok := m.hosts[host]; ok {
+		found = true
+		md = old
+	}
+
+	md.InternalHostname = host
+	md.Meta.Hostname = host
+
+	// InstanceID field
+	if instanceId, ok, fieldErr := getInstanceID(res.Attributes()); fieldErr != nil {
+		err = multierr.Append(err, fieldErr)
+	} else if ok {
+		old := md.Meta.InstanceID
+		changed = changed || old != instanceId
+		md.Meta.InstanceID = instanceId
+	}
+
+	// EC2Hostname field
+	if EC2Hostname, ok, fieldErr := getEC2Hostname(res.Attributes()); fieldErr != nil {
+		err = multierr.Append(err, fieldErr)
+	} else if ok {
+		old := md.Meta.EC2Hostname
+		changed = changed || old != EC2Hostname
+		md.Meta.EC2Hostname = EC2Hostname
+	}
+
+	// Gohai - Platform
+	md.Gohai.Gohai.Platform.(map[string]string)["hostname"] = host
```

Comment:
I thought you agreed to add a method for accessing this map to avoid this ugliness :)

Reply:
I was thinking a getter/setter, e.g.:
```go
md.SetPlatform(k, v)
```
But the way you did it is fine too. Except you aren't using it here.

Reply:
oops, using it now